src/main/scala/com/fulcrumgenomics/bam/ComputeRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,330 @@
+
+    // Check for overlapping reads
+    if (CoordMath.overlaps(rec.getAlignmentStart, rec.getAlignmentEnd, rec.getMateAlignmentStart, mateAlignmentEnd)) {
+      fail(s"Record overlaps with its mate, consider using ClipOverlappingReads.  Record: $rec")
+    }
+
+    // Get the overlapping genes
+    val leftMostBase     = Math.min(rec.getAlignmentStart, rec.getMateAlignmentStart)
+    val rightMostBase    = Math.max(rec.getAlignmentEnd, mateAlignmentEnd)
+    val recInterval      = new Interval(rec.getReferenceName, leftMostBase, rightMostBase)
+    val overlappingGenes = overlapDetector.getOverlaps(recInterval)
+
+    if (overlappingGenes.size() == 1) {
+      val mateAlignmentBlocks = SAMUtils.getAlignmentBlocks(mateCigar, rec.getMateAlignmentStart, "mate cigar")
+      val (lowerReadAlignmentBlocks, upperReadAlignmentBlocks) = if (rec.getAlignmentStart < rec.getAlignmentEnd) {
+        (rec.getAlignmentBlocks, mateAlignmentBlocks)
+      }
+      else {
+        (mateAlignmentBlocks, rec.getAlignmentBlocks)
+      }
+
+      val gene = overlappingGenes.head
+      calculateInsertSizeFromRemainingTranscripts(
+        gene                     = gene,
+        lowerReadAlignmentBlocks = lowerReadAlignmentBlocks,
+        upperReadAlignmentBlocks = upperReadAlignmentBlocks,
+        remainingTranscripts     = gene.toIterator
+      )
+    }
+    else {
+      None
+    }
+  }
+
+  /**
+    * Calculates the insert size for all remaining transcripts ensuring that the results are the same.  If a transcript
+    * is encountered that yields an insert size different than previously calculated, computation is stopped, and None
+    * is returned.
+    */
+  @tailrec
+  private def calculateInsertSizeFromRemainingTranscripts(gene: Gene,
+                                                          lowerReadAlignmentBlocks: Iterable[AlignmentBlock],
+                                                          upperReadAlignmentBlocks: Iterable[AlignmentBlock],
+                                                          remainingTranscripts: Iterator[Transcript],
+                                                          result: Option[Int] = None): Option[Int] = {
+    if (remainingTranscripts.isEmpty) { // no more transcripts, return the result
+      result
+    }
+    else {
+      // calculate the result for this transcript
+      val transcript = remainingTranscripts.next()
+      val exons      = transcript.toIterator.bufferBetter
+
+      // calculate the # of bases overlapping the lower read
+      val lower = numReadBasesOverlappingTranscript(lowerReadAlignmentBlocks.toIterator.bufferBetter, exons)
+
+      // calculate the # of bases overlapping the upper read
+      val middle = {
+        val start = upperReadAlignmentBlocks.head.getReferenceStart // don't move past the start of the upper blocks
+        exons.takeWhile(_.end < start).map { e => e.end - e.start + 1 }.sum
```

Comment:
This confused me because it wasn't clear until I got there, that `numReadBasesOverlappingTranscript` consumed out of the `exons` iterator and left it in the right place.